yoti-sdk-impl/src/main/java/com/yoti/api/client/spi/remote/util/AnchorCertificateUtils.java

```diff
@@ -0,0 +1,111 @@
+package com.yoti.api.client.spi.remote.util;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import org.bouncycastle.asn1.ASN1InputStream;
+import org.bouncycastle.asn1.ASN1OctetString;
+import org.bouncycastle.asn1.ASN1Primitive;
+import org.bouncycastle.asn1.ASN1TaggedObject;
+import org.bouncycastle.asn1.DEROctetString;
+import org.bouncycastle.asn1.DLSequence;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.protobuf.ByteString;
+import com.yoti.api.client.spi.remote.proto.AttrProto.Anchor;
+
+public class AnchorCertificateUtils {
+    private static final Logger LOG = LoggerFactory.getLogger(AnchorCertificateUtils.class);
+
+   
+    public static AnchorVerifierSourceData getTypesFromAnchor(Anchor anchor) {
+        Set<String> types = new HashSet<String>(anchor.getOriginServerCertsCount());
+        AnchorType anchorType = AnchorType.UNKNOWN;
+
+        try {
+            CertificateFactory factory = CertificateFactory.getInstance("X.509");
+
+            for (ByteString bs : anchor.getOriginServerCertsList()) {
+                X509Certificate certificate = (X509Certificate) factory.generateCertificate(new ByteArrayInputStream(bs.toByteArray()));
+
+                List<String> extensions = new ArrayList<String>();
+                for (AnchorType type : AnchorType.values()) {
+                    extensions = getListOfStringFromExtension(certificate, type.extensionOid);
+                    if (extensions.size() > 0) {
+                        anchorType = type;
```

Comment:
We iterate over each __AnchorType__ in each __ByteString__, updating the value of __anchorType__ each time we find one with extensions.  Then at the end we just use the __last value__ we encountered.

So, are we only expecting one match?  Might be easier to follow if this had a unit test dedicated to it.

And I agree with @DarthSeppius  - this class only has one public method, it doesn't need to be called "xxxUtils"

Reply:
The code is iterating over a certificate chain. Only one of the elements in the chain will be an actual anchor with a value. 
`getListOfStringFromExtension` verifies that by looking for a specific objectId (one for SOURCE and one for VERIFIER ) and, if found, it returns the corresponding object value as a set of strings.

The alternative would be to just check for the type and have another method that actually parses the object, but I don't really see much value in it.

It is thus inevitable to go through all the elements until a matching one is found.

In terms of tests, this is covered in `SimpleProfileTest`.

Finally, I agree this shouldn't be a named as a util. It is indeed a parser, so I will go with `AnchorCertificateParser`.

Reply:
Ok, so it's doing a find-first.  To be fair I should have spotted the __break;__ statement